### PR TITLE
653 - Integrate Cypress to run automatically on Github

### LINKED ISSUE #653: Integrate Cypress to run automatically on Github

### PR DESCRIPTION
**Pull request by narendrasss**
Closes #653

## What?

This PR introduces a Github Actions workflow to run Cypress automatically on every PR. It does this using Cypress's github actions workflow + the docker changes introduced in #706.

## Notes

- I'm running into some permissions issues so that H5P and GF cannot be installed. A workaround to this would be to commit the plugin folders themselves (not ideal).
- Not all of the tests are passing. There seems to be some bugs on master causing some of the tests to fail (a bunch of console errors when the Tapestry is empty):

Screen Shot 2020-08-24 at 14 12 35

(Naturally any tests related to H5P and GF would fail as well)

### CODE REVIEW

**Comment by wynnset** [comment]
Please merge in #701 and do some more testing and then this is good to get merged.

**Comment by narendrasss** [comment]
Todo: Look into why WP CLI isn't installed correctly.

**Comment by wynnset** [comment]
Seems like tests are not passing now. @rlnsy do you think you could take a look?

**Comment by narendrasss** [comment]
@rlnsy if you decide to take a look at this, just getting the tests to _run_ is good enough for this PR to get merged in. Don't worry about the tests passing.

**Comment by rlnsy** [comment]
@narendrasss Looking at the test logs, at least the WP CLI issues should be resolved by my PR (#753)!. Perhaps we can wait for that to get test/merged in then rebase?

**Comment by narendrasss** [comment]
@rlnsy Sounds good to me! I'll label this as blocked until #753 is tested.

**Comment by wynnset** [comment]
@rlnsy / @narendrasss this is now unblocked since #753 was merged in.

**Comment by rlnsy** [comment]
@narendrasss Looks like we are just waiting on #779 for this. In the meantime I'll create a mock branch and test it out.

**Comment by rlnsy** [comment]
@narendrasss This GitHub runner is currently producing these results consistently:
Screen Shot 2020-10-15 at 10 01 31 AM
but I'm getting a different fail-rate running locally (like 23%). My goal is to make these consistent with each other before merging. I'm wondering if it has to do with modifying the testing configuration, like I noticed you changed the test script to
```
"test:e2e": "cypress run --headless -b chrome --config watchForFileChanges=false"
```
Can you let me know what results you're getting locally?

**Comment by narendrasss** [comment]
@rlnsy That looks good, except all my tests on `settings.spec.js` pass. I'm rerunning the tests right now locally and I'll update you.

**Comment by rlnsy** [comment]
@narendrasss Great! Yeah same for me.

**Comment by narendrasss** [comment]
@rlnsy Getting these results locally:

Screen Shot 2020-10-15 at 14 14 50

There are a couple issues with test flake that might be causing some tests to fail when it shouldn't. That's a bit out of scope of this PR though.

**Comment by rlnsy** [comment]
@narendrasss Yeah I get the same thing locally! Looks like 3-4 of the tests are consistently failing in the Github environment - I'm looking more into the actual output to see what's going on - perhaps a quirk of the cypress runner?

**Comment by narendrasss** [comment]
@rlnsy I updated the timeouts but unfortunately it didn't look like it solved the problem. One thing that might help us is getting access to the videos recorded by the runs, but it seems we need to set up Cypress dashboard for that to work.

@wynnset do you think you can look into integrating the dashboard? We would need to set up an account then link the account to this repo using the github app. Some useful links:

- [About the dashboard](https://docs.cypress.io/guides/dashboard/introduction.html#Features)
- [On integrating the dashboard to github](https://docs.cypress.io/guides/dashboard/github-integration.html#Install-the-Cypress-GitHub-app)
- [The github app](https://github.com/apps/cypress)

**Comment by wynnset** [comment]
Hi @narendrasss I set this up. I also set up Github secrets for CYPRESS_PROJECT_ID and CYPRESS_RECORD_KEY which are needed to run cypress tests on the Test Runner. I also added you and @rlnsy as users in the organization.

**Comment by narendrasss** [comment]
Thanks for setting this up @wynnset! @rlnsy looks like there might be some permission issues which causes some of the tests to fail. I'll assign this ticket back to you.

**Comment by wynnset** [comment]
We'll merge with the errors and create another PR for dealing with fixing those specific upload-related errors.

**Comment by wynnset** [comment]
@narendrasss / @rlnsy could one of you please fix the merge conflicts and then assign it back to me for review?

**Comment by rlnsy** [comment]
@narendrasss Noticed that you merged master in - did things break to warrant the needs work label?

**Comment by narendrasss** [comment]
@rlnsy Yes - quite a few tests are failing that was ok before. I fixed a few but there are 3 left that needs to be worked on.

**Comment by narendrasss** [comment]
Finally good to go now @rlnsy @wynnset 😄 

I updated some of the test specs with labels if they are expected to fail in CI/is a known bug. Also reduced a bit of flakiness. Apologies for the scope creep of this PR - I figured it would be good to handle it all at once.

**Comment by wynnset** [comment]
Sorry seems like a few things are failing after the merge.

File: templates/vue/cypress/fixtures/three-nodes.json
Comment (wynnset) [inline templates/vue/cypress/fixtures/three-nodes.json:244]: Is it possible to move this somewhere else so we don't have to repeat it in every fixture?

Comment (narendrasss) [reply templates/vue/cypress/fixtures/three-nodes.json:244]: I had to add this to every fixture because #682 requires imported Tapestry JSON files to contain the property. It was only missing because the tapestries were made prior to #682 - we wouldn't have to add it if we use newer Tapestries.

I might know a way to modify the Cypress code so this may not be necessary, but would require changing every single spec file that uses `cy.setup` - not sure if that's worthwhile.

Comment (wynnset) [reply templates/vue/cypress/fixtures/three-nodes.json:244]: Ok I see, that's fine then. We can leave it as-is for now.

**Review by wynnset — APPROVED** [review]

**Pull request merged**
